WCAG Updates - 2020

jakerarr opened the pull request:
- Add ability to pass heading level to FormConfig, with a default of H3
- Update labels, buttons, and spans in fileUpload for proper conformity to AA standards
- Add SR only text to select all/deselect all buttons in selectMulti
- Remove labels from statics (no actual input), change to strong text
- Update toggleTrueFalse and toggleYesNo to use formgroups and legends, remove label from div wrapper

jakerarr committed:
Update textAreaInput.twig

jakerarr committed:
Update checkBox.twig

jakerarr committed:
Update checkBoxCentered.twig

Commit:
WCAG update for form controller

Commit:
WCAG update for form controller

Commit:
WCAG update for form controller

Commit:
WCAG update for form controller

Commit:
WCAG update for form controller

Commit:
WCAG update for form controller

Commit:
WCAG update for form controller

Commit:
moving setTableId to getTableId in FormArrayElement in the cases where table_id is empty

Commit:
moving setTableId to getTableId in FormArrayElement in the cases where table_id is empty

Commit:
adding specificity in table_id generation, in-lining aria-labelledby logic

Commit:
updating CHANGELOG.md

Commit:
fixing issue with double double quotes in field aria-labelledby, adding ID to ten scale NA

Commit:
fixing issue with double double quotes in field aria-labelledby, adding ID to ten scale NA

Commit:
fixing issue with double double quotes in field aria-labelledby, adding ID to ten scale NA

Commit:
Merge remote-tracking branch 'KongHack/master'

- Update to Mitch's latest version before making changes

Commit:
- Add changes from upstream

Commit:
- Merge remote-tracking branch 'KongHack/master'

Commit:
- WCAG updates
- Fix labels on statics and toggles

Commit:
- Merge remote-tracking branch 'KongHack/master'
- Handle merge conflicts, commit merge

Commit:
- Fix bug in toggleYesNo

GameCharmer commented on src/Forms/FormConfig.php:
```diff
@@ -18,6 +18,8 @@ class FormConfig implements FieldContainerInterface
     const OVERRIDE_SUBMIT      = 'submitButton';
     const OVERRIDE_PANEL_CLASS = 'panelClass';
 
+    const DEFAULT_NAVIGATION_HEADING_LEVEL = 'h3';
```
Leave the default div, override in local version of FormConfig

jakerarr replied:
Got it!

GameCharmer commented on src/Forms/FormConfig.php:
```diff
@@ -193,6 +196,35 @@ public function getFormId()
+     * @param $value
+     * @return $this
+     */
+    public function setNavigationHeadingLevel($value)
```
This limits you to H1-6.  Since this is a library, don't do that.

jakerarr replied:
Done!

GameCharmer commented on src/Forms/FormField.php:
```diff
@@ -91,9 +91,9 @@ class FormField
-
     /**
      * @var string
+
```
This random empty line will break docblock processing

jakerarr replied:
Removed!

GameCharmer commented on twig/BS3/fields/fileInput.twig:
```diff
@@ -1,14 +1,15 @@
+            </button>
+        </span>
+        <label class="sr-only" for="{{ field.id }}_path">File Path</label>
+        <input id="{{ field.id }}_path" type="text" class="form-control" readonly {{ field.getDataAttributeString()|raw }}>
```
What does this second input do?  Realistically, we should only have the file input here.

jakerarr replied:
This input existed, I just added the label.  I believe it's just to show the file path.  Let me know if you want it removed!

GameCharmer replied:
Odd that it shows as an entirely new line.  Leave as is.

GameCharmer commented on twig/BS3/fields/selectMultipleInput.twig:
```diff
@@ -23,13 +23,13 @@
         <span class="input-group-btn">
-            <button class="btn btn-success tool_button" id="{{ id }}-all" type="button" title="Select All">
-                <i class="fa fa-check"></i>
+            <button class="btn btn-success tool_button" id="{{ id }}-all" type="button">
```
We want the title on this so people can hover over and know what it does.  This control isn't just for screen readers.

jakerarr replied:
Got it!

GameCharmer commented on twig/BS3/fields/selectMultipleInput.twig:
```diff
@@ -23,13 +23,13 @@
-            <button class="btn btn-danger tool_button" id="{{ id }}-none" type="button" title="Remove All">
-                <i class="fa fa-times"></i>
+            <button class="btn btn-danger tool_button" id="{{ id }}-none" type="button">
+                <i class="fa fa-times"></i><span class="sr-only">Select All</span>
```
We want the title on this so people can hover over and know what it does.  This control isn't just for screen readers.

jakerarr replied:
Got it!

GameCharmer commented on twig/BS3/fields/static.twig:
```diff
@@ -1,4 +1,23 @@
 {% macro render(field) %}
+    {% if field.label is not null %}
```
What is all of this for?  Static fields are display only.

jakerarr replied:
Static fields don't have an input, and so the label is an "orphaned" label.  This is to make it so that they just get a strong text instead of a label. Visually stays the same, for screen readers it doesn't announce an input that doesn't exist.

GameCharmer replied:
Is there a universal label that can apply to both an input and a div?  You can add `contenteditable="true"` to a div and the content becomes editable, and it's not an input, select, nor textarea.

jakerarr replied:
cool!

GameCharmer commented on twig/BS3/fields/toggleTrueFalse.twig:
```diff
@@ -1,14 +1,25 @@
-    <div>
-        <div class="btn-group" id="{{ field.id }}" data-toggle="buttons">
+    <fieldset>
+        {% if field.label is not null %}
```
These inline overrides like this cause duplication.  We most likely don't want every heading twice.

GameCharmer commented on twig/BS3/fields/toggleYesNo.twig:
```diff
@@ -1,14 +1,25 @@
 {% macro render(field) %}
-    <div>
-        <div class="btn-group" id="{{ field.id }}" data-toggle="buttons">
+    <fieldset>
```
These inline overrides like this cause duplication. We most likely don't want every heading twice.

GameCharmer commented on twig/BS3/fields/toggleYesNo.twig:
```diff
@@ -1,14 +1,25 @@
             </label>
-            <label class="btn btn-default no {{ (field.value=='N' ? 'active' : '') }}">
-                <input type="radio" class="{{ field.class }}" name="{{ field.name }}" value="N" autocomplete="off" {{ (field.value=='N' ? 'checked="checked"' : '') }} {{ field.getDataAttributeString()|raw }}> No
+            <label class="btn btn-default no {{ (field.value==0 'N' ? 'active' : '') }}">
```
`field.value==0 'N' ` is just straight up broken

jakerarr replied:
Gah... I think I overwrote the change on this one.  Yeah- it's fixed.

GameCharmer commented on twig/BS3/forms/controller.twig:
```diff
@@ -13,13 +13,12 @@
             <div class="panel panel-sitespecific">
                 <div class="panel-heading">
-                    <div class="panel-title">Navigation</div>
+                    <{{ navHeading }} class="panel-title">Navigation</{{ navHeading }}>
```
use `FC_Config.getNavHeading()` (or whatever the function name is)

jakerarr replied:
Changed!

GameCharmer commented on twig/BS3/forms/controller_block.twig:
```diff
@@ -13,7 +13,7 @@
     <div class="panel panel-{{ twigOverrides.panelClass|default('sitespecific') }}">
         <div class="panel-heading">
-            <div class="panel-title">{{ forms[activeForm]|raw }}</div>
+            <h3 class="panel-title">{{ forms[activeForm]|raw }}</h3>
```
Don't you want your nav heading?

jakerarr replied:
Switched this to use the nav heading

GameCharmer commented on twig/BS3/macros/forms.twig:
```diff
@@ -38,23 +38,26 @@
-                            {% if field.reqLevel > 1 %}
-                                {{ myself.requiredIndicator(field) }}
-                            {% endif %}
+                    {% set labelIdSwitch = {0:'TYPE_TOGGLE_TRUE_FALSE', 1:'TYPE_TOGGLE_YES_NO', 2:'TYPE_STATIC'} %}
```
Sorry, this is a straight up no.  We're not coding in static arrays like this.  Since this is a library, anyone can add their own fields.  See src/FormFields in our own project.

jakerarr replied:
Should I add this to the definitions (like getUseStandardLabel() or something)?  The big problem that we have is labeling divs- for WCAG stuff, labels are only for form elements like input, select, etc.  For things like toggles that require a field set, the standard labeling in forms.twig needs that little tweak.  I'm open to the best way to do this!

GameCharmer replied:
Would it be better to use fieldsets with legends instead of divs with labels everywhere?

jakerarr replied:
It's best to use them for multi-input type things (like radios and check boxes) or sets of data (address).  According to WCAG standards, the fieldset and legend should be used to describe a set of inputs that relate to each other in some way.  Each input in a fieldset should have its own label.  The legend should describe the group of inputs.  In the ACEMAPP side of things, I also added a fieldset around the address stuff- it's all address stuff, and each item is individually labeled (as city, state, street etc) even though the labels are sr-only.

I think adding a fieldset around each input would make things worse- they would hear the fieldset legend, then the label, which would probably be the same thing :\

GameCharmer replied:
Hence the `instead of` on that last bit.  I'm suggesting removing the label and running with fieldset/legend only instead of div/label.

When it comes to reading, is there any difference between reading the fieldset legend vs reading the div label?

GameCharmer replied:
Discussed in person.   Mark this one resolved when complete.

GameCharmer reviewed (commented):
Please review requested changes

Commit:
- WCAG updates
- Patch issue with standard labels
- Patch issue with file button submitting form

Commit:
Merge remote-tracking branch 'KongHack/master'
- Merge master

Commit:
- Last changes

GameCharmer committed:
Update FormConfig.php

GameCharmer committed:
Update FormField.php
